File: internships/yandex/shad/cpp/3.cpp
#include <bits/stdc++.h>
using namespace std;

const int N = 500 + 10;
const long long INF = 1e18;
int n;
long long dist[N][N];
void solve()
{
    cin >> n;
    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= n; j++)
        {
            cin >> dist[i][j];
            if (dist[i][j] == -1)
            {
                dist[i][j] = INF;
            }
        }
    }
    for (int k = 1; k <= n; k++)
    {
        for (int i = 1; i <= n; i++)
        {
            for (int j = 1; j <= n; j++)
            {
                dist[i][j] = max(dist[i][j], min(dist[i][k], dist[k][j]));
            }
        }
    }
    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= n; j++)
        {
            if (dist[i][j] == INF || i == j)
            {
                cout << 2000000023 << " ";
            }
            else
            {
                cout << dist[i][j] << " ";
            }
        }
        cout << endl;
    }
}

int main()
{
    ios::sync_with_stdio(NULL), cin.tie(0), cout.tie(0);
    solve();
}

File: internships/yandex/shad/1/python/3.py
n = int(input())
g = [list(map(int, input().split())) for _ in range(n)]

for i in range(n):
    for j in range(n):
        if g[i][j] == -1:
            g[i][j] = 10**18

for k in range(n):
    for i in range(n):
        for j in range(n):
            g[i][j] = max(g[i][j], min(g[i][k], g[k][j]))


for i in range(n):
    for j in range(n):
        if i == j or g[i][j] == 10**18:
            g[i][j] = 2000000023


for i in range(n):
    print(*g[i])
